software/react-native/concepts.md

```diff
@@ -0,0 +1,60 @@
+# Basic Concepts
+
+#### Request / Response Data Flow
+
+In a standard redux project data flows like this:
+
+1. Dispatch an `attempting` action that will attempt to `POST`/`GET` data from a server. This action will usually flip an `attempting: true` flag in it's domain-related reducer in case we want to display some sort of UI to demonstrate to the user that a request is in process.
+
+2. A saga is listens for this `attempting` action, and then runs when it receives it along with it's params (*it will have passed through the reducer first*). The saga handles all side-effects associated with this action. The saga will be the place where the api call is actually made. Depending on it's success or failure a `success` or `failure` action will be dispatched to the reducer. If it's a `success` we dispatch the received data as well, if it's a `failure`, we display the received error messages.
+
+3. After the `success` action is dispatched from the reducer, the data will get stored as is or [normalized](https://github.com/paularmstrong/normalizr) and stored. We try to normalize all deeply nested data to make it much easier to work with. The `success` or `failure` action will also flip that `attempting: true` flag mentioned above back to `attempting: false` to show the user that the request has completed.
+
+4. Now that we have attempted to fetch our data, successfully received our data, and stored (or normalized) it, depending on how we are using it, we may need to *transform* the data before displaying or performing calculations with it.
+In a perfect world, we would not need to do many transformations since the API would return to us the data exactly as we needed *cough GraphQL*. However, often transforms are necessary to coerce the data into a more workable form. For example, List Views in React Native require very specific data structures, and we often have to transform our data to fit a List View's requirements.
+
+5. The cycle is now complete. To recap: we first dispatch a redux action which may trigger an attempting flag. The saga that is listening for this action receives it and any params passed in and makes the actual API call. Regardless of success or failure, the saga will dispatch an action back to the reducer with the result of the API call. If it was a success we may normalize the data before we store it. If it was a failure, we may display the error messages. Finally, depending on how we want to use the data, we may transform it before using it in the UI.
+
+
+
+#### Redux Saga
+
+[Redux Sagas](https://github.com/yelouafi/redux-saga) are middleware for handling side-effects in Redux.
+
```

Comment:
also good to expand on what 'side effects' are, what are some possible things you can use it for other than API calls? (ie. dispatching other actions like putting up a modal screen while a fetch is happening).

One other essential reason for sagas is to help make the async environment of JS a bit more bearable when dealing with things like APIs since it helps to coordinate all the different APIs in tandem.

With sagas, we can allow certain function calls to be 'blocking' (and therefore synchronous ie. `call`) so that it is more predictable, but we could also dispatch async calls (ie. `put`) as well